add KMS key to profile and function config.

DonMcNamara opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:

Add KMS key configuration to function and provider configuration.

Closes #2996

<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:

Added ability to set the Lambda KMS to the provider and function configuration, using configuration key `kmsKeyArn`.

<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

## How can we verify it:

<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* AWS CLI commands - To list AWS resources and show that the correct config is in place
* Other - Anything else that comes to mind to help us evaluate
-->
### 1. Create test service
```
sls create --template aws-nodejs --path kms-test-service
```

### 2. Deploy service
```
cd kms-test-service
sls deploy --stage dev
```
### 3. Describe the service to see the KMS key used
```
aws lambda get-function-configuration --function-name kms-test-service-dev-hello
```
**You should not see a KMSKeyArn value at this point.** A missing KMSKeyArn means you are using default service key. [See the CLI docs.](http://docs.aws.amazon.com/cli/latest/reference/lambda/get-function-configuration.html)
You can also check in the aws console (lambda function, configuration tab, advanced settings toggle, at the very bottom)

### 4. Choose a KMS key to use.
```
aws kms list-keys
```

### 5. Edit your serverless.yml file to include the KMS key ARN, for example:
```
provider:
  name: aws
  runtime: nodejs4.3
  kmsKeyArn: [YOUR KEY ARN, something like arn:aws:kms:us-east-1:XXXXXXX:key/key-id]
```

### 6. Redeploy your function
```
serverless deploy --stage dev
```

### 7. Describe your service again. 
```
aws lambda get-function-configuration --function-name kms-test-service-dev-hello
```
**You should now see a KMSKeyArn value using the key you specified in serverless.yml**

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Enable ["Allow edits from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) for this PR
- [x] Change ready for review message below


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

pmuens commented on docs/providers/aws/guide/functions.md:
```diff
@@ -253,6 +253,29 @@ functions:
+service: service-name
+provider:
+  name: aws
+  kmsKeyArn: [your KMS key ARN]
```
Would suggest to add an `environment` variable here because the KMS key only works together with an environment variable

DonMcNamara replied:
Addressed in https://github.com/serverless/serverless/pull/2998/commits/8fe133283ef6aacf1c45cb4a678f9bccb1c362b3

pmuens commented on docs/providers/aws/guide/functions.md:
```diff
@@ -253,6 +253,29 @@ functions:
+    handler: handler.hello
+  users: # this function will OVERWRITE the service level environment config above
+    handler: handler.users
+    kmsKeyArn: [your KMS key ARN]
```
Would suggest to add an `environment` variable here because the KMS key only works together with an environment variable

DonMcNamara replied:
Addressed in https://github.com/serverless/serverless/pull/2998/commits/8fe133283ef6aacf1c45cb4a678f9bccb1c362b3

DonMcNamara commented:
Thanks for the review @pmuens 

I've addressed your comments on the changes.

Also, I had a slight misunderstanding with how KMS worked with lambda environment variables.

The KMS key specified on the lambda is used to encrypt/decrypt your environment variables **at rest**. This is enabled for all environment variables and is transparent to the user. The values are decrypted before being set for your lambda's environment.

When your environment variables include secrets, AWS recommends, in addition to the at-rest encryption, that you encrypt your values. In this case, there are two layers of encryption. You then decrypt the secrets in your lambda code. You can use KMS to do this additional layer of encryption/decryption and you can use any key that your lambda (execution IAM role) has permission to use.

I had originally thought you had to use the KMS key specified on the lambda for the second layer of encryption, but that isn't true.

Let me know what you think!

DonMcNamara commented:
Also, I added some steps to test to the PR comment.

jmpage commented:
@mthenw are you looking for this modification to happen during an `sls deploy`?

DonMcNamara commented:
@mthenw Thanks for the review.

I'm not going to have time to work on the IAM change. I'm going to close this PR for now, but If anyone wants to move it forward, feel free to re-open it.

pmuens commented:
@DonMcNamara thanks for your hard work on this PR! 👍 💯 

We keep it in the pipeline and might reopen once it's on our plates again!
